### ISSUE TITLE
functional count does not flatten as expected

### ISSUE THREAD

**Issue by GustavBjordal** [MiniZinc/libminizinc#427]
Under MiniZinc 2.5.1 and 2.5.2 and Gecode 6.3.0, the following is not properly flattened.
```
array[1..20] of var 0..6: X;
constraint count(X,0)=5;
```
becomes
```
array [1..20] of int: X_INTRODUCED_62_ = [-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1];
var 0..6: X_INTRODUCED_0_;
...
var bool: X_INTRODUCED_20_ ::var_is_introduced :: is_defined_var;
var 0..1: X_INTRODUCED_21_ ::var_is_introduced :: is_defined_var;
...
constraint int_lin_eq(X_INTRODUCED_62_,X_INTRODUCED_60_,-5);
constraint int_eq_reif(X_INTRODUCED_0_,0,X_INTRODUCED_20_):: defines_var(X_INTRODUCED_20_);
constraint bool2int(X_INTRODUCED_20_,X_INTRODUCED_21_):: defines_var(X_INTRODUCED_21_);
...
```
rather than the expected 
```
predicate fzn_count_eq(array [int] of var int: x,var int: y,var int: c);
var 0..6: X_INTRODUCED_0_;
...
constraint fzn_count_eq(X,0,5);
```

This seems to be caused by some internal rewrite that the compiler does of `function var int: count(array[$X] of var int: x, var int: y)`.

In `count_fn.mzn` (in the std folder) if I remove the `count(array[$X] of var int: x, var int: y)` function completely, then the model no longer compiles (as you would expect).
However, if I change the file to:
```
function var int: count(array[$X] of var int: x, var int: y) ::promise_total =
  123;
%  let { var 0..length(x): c::is_defined_var; constraint count(array1d(x),y,c)::defines_var(c); } in c;
```
Then not only does the model still compile, it still produces the same flatzinc!

So clearly the definition of `count` is ignored by the compiler and replaced by something else somewhere.
This seems to be what is preventing the compiler from reaching the expected `fzn_count_eq(X,0,5);` flattening.

**Comment by guidotack** [comment]
The compiler internally rewrites `count(x,y)=z` into `count_eq(x,y,z)`, bypassing the functional version. In this case, since `y` is par, it will then rewrite into `fzn_count_eq_par`, which has no redefinition in Gecode. We should probably change this so that `fzn_count_eq_par` is rewritten into `fzn_count_eq` (which has an identical decomposition, so it's not going to change anything for solvers that don't redefine it).

**Comment by GustavBjordal** [comment]
Ok the commit seems to take care of it, thanks!

However, isn't is a bit misleading that `count(x,y)=z` has a definition in `count_fn.mzn` even when it is ignored by the compiler?
Wouldn't it make more sense to just declare it without a definition:
```
function var int: count(array[$X] of var int: x, var int: y) ::promise_total;
```
I found it a bit weird that when I added tracing to the definition it was ignored and never printed. But I understand why now.
